core-primitives/stf-state-handler/src/state_handler.rs

```diff
@@ -25,144 +25,261 @@ use crate::{
 	error::{Error, Result},
 	handle_state::HandleState,
 	query_shard_state::QueryShardState,
+	state_initializer::InitializeState,
 	state_snapshot_repository::VersionedStateAccess,
 };
+use itp_hashing::Hash;
+use itp_sgx_externalities::SgxExternalitiesTrait;
 use itp_stf_state_observer::traits::UpdateState;
 use itp_types::ShardIdentifier;
-use std::{sync::Arc, vec::Vec};
+use std::{collections::HashMap, sync::Arc, vec::Vec};
+
+type StatesMap<State, Hash> = HashMap<ShardIdentifier, (State, Hash)>;
 
 /// Implementation of the `HandleState` trait.
 ///
-/// It's a concurrency wrapper around a state snapshot repository, which handles
-/// access to any shards and state files. The state handler ensures we have thread-safe
-/// concurrent access to that repository.
-pub struct StateHandler<Repository, StateObserver> {
+/// Responsible for handling any state instances. Holds a map with all the latest states for each shard.
+/// In addition, uses the snapshot repository to save file snapshots of a state.
+pub struct StateHandler<Repository, StateObserver, StateInitializer>
+where
+	Repository: VersionedStateAccess,
+{
 	state_snapshot_repository: RwLock<Repository>,
+	states_map_lock: RwLock<StatesMap<Repository::StateType, Repository::HashType>>,
 	state_observer: Arc<StateObserver>,
+	state_initializer: Arc<StateInitializer>,
```

Comment:
This is the heart of this PR: The state handler itself now holds a map of states. The snapshot repository is only used to write snapshots to file, but read operations don't have to first read and decrypt a file, but can directly access the in-memory map here.